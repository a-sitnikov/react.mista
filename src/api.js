//@flow
import fetchJsonp from 'fetch-jsonp'
import * as utils from './utils'

export const domain = localStorage.getItem('domain') || 'https://forum.mista.ru';
//export const domain = 'https://dev.mista.ru';
export const urlTopicsList    = localStorage.getItem('urlTopicsList') || 'api/topic.php';
export const urlTopicInfo     = localStorage.getItem('urlTopicInfo') || 'ajax_gettopic.php';
export const urlTopicMessages = localStorage.getItem('urlTopicMessages') || 'api/message.php';
export const urlLogin         = localStorage.getItem('urlLogin') || 'ajax_login.php';
export const urlLogout        = localStorage.getItem('urlLogout') || 'users.php'; // users.php?&action=exit
export const urlCookies       = localStorage.getItem('urlCookies') || 'ajax_cookie.php';
export const urlSections      = localStorage.getItem('urlSections') || 'ajax_getsectionslist.php';
export const urlNewMessage    = localStorage.getItem('urlNewMessage') || 'topic.php?id=:id';
export const urlNewTopic      = localStorage.getItem('urlNewTopic') || 'index.php';
export const urlAddBookmark   = localStorage.getItem('urlAddBookmark') || 'ajax_addbookmark.php';
export const urlSearch        = localStorage.getItem('urlSearch') || 'ajax_find.php';

// Topics list
export type RequestTopicsList = {
    topics?:   number, // число тем, которые нужно вернуть (по умолчанию - 20);
    utime?:    number, // время (в формате unixtime) после которого надо вернуть ветки;
    callback?: string, // если указано, то ответ заворачивается в вызов функции, имя которой передано в параметре (реализация JSONP)
    forum?:    string, // раздел форума (1c|it|life);
    section_short_name?: string,
    user_id?:  string,
    mytopics?: string
}

export type ResponseTopicsListItem = {
    id: number,
    forum: string,
    sect1: string,
    sect2: string,
    v8: string,
    closed: number,
    down: number,
    text: string,
    utime: number,
    created: number,
    user: string,
    user0: string,
    is_voting: number,
    answ: number
}

export type ResponseTopicsList = Array<ResponseTopicsListItem>;

export const getTopicsList = async (params: RequestTopicsList): Promise<ResponseTopicsList> =>  {
    const json = await fetchJsonpAndGetJson(urlTopicsList, params);
    return json;
} 


// Topics info
export type RequestInfo = {
    id: string // идентификатор (topic_id) темы
}

export type ResponseVoteItem = {
    select: string,
    result: number
}

export type ResponseInfo = {
    id: string,
    text: string,
    forum?: string,
    section?: string,
    created?: string,
    user_id?: string,
    user_name?: string,
    answers_count: string,
    down?: number,
    closed?: number,
    deleted?: number,
    is_voting?: number,
    voting?: Array<ResponseVoteItem>
}

export const defaultInfo = {
    id: "",
    text: "",
    answers_count: ""
}

export const getTopicInfo = async (params: RequestInfo): Promise<ResponseInfo> =>  {
    const json = await fetchJsonpAndGetJson(urlTopicInfo, params);
    return json;
} 


// Topic messages
export type RequestMessages = {
    id:    number | string, // идентификатор (topic_id) темы
    from?: number, // с какого сообщения. если не указан, то с первого
    to?:   number  // до какого сообщения. если не указан, то from+10
}

export type ResponseMessage = {
    id:    string,
    n:     string,
    user:  string,
    userId: string,
    text:  string,
    utime: string,
    vote:  number
}

export type ResponseMessages = Array<ResponseMessage>;

export const getTopicMessages = async (params: RequestMessages): Promise<ResponseMessages> =>  {
    const json = await fetchJsonpAndGetJson(urlTopicMessages, params);
    return json;
} 

export const getMessage = async (id: number | string, n: number): Promise<?ResponseMessage> =>  {
    const json = await getTopicMessages({
        id, 
        from: n, 
        to: n+1}
        );

    let data;
    if (json && json.length > 0)
        data = json.find(val => val.n === String(n));

    return data;
} 

export const getTopicMessagesCount = async (id: number | string): number => {
    const info = await getTopicInfo({id: String(id)});
    return +info.answers_count;
}

// Login
export type RequestLogin = {
    username:  string, // логин (ник) пользователя
    password:  string, // пароль пользователя
    callback?: string, // если указано, то ответ заворачивается в вызов функции, имя которой передано в параметре (реализация JSONP)
}

export type ResponseLogin = {
    userid: string,
    username: string,
    hashkey: string,
    error: string
}

export const getLogin = async (params: RequestLogin): Promise<ResponseLogin> =>  {
    const json = await fetchJsonpAndGetJson(urlLogin, params);
    return json;
} 

export const getLogout = async(): Promise<any> => {
    const params = { action: "exit" };
    await fetchJsonpAndGetJson(urlLogout, params);
}


// Cookies
export type ResponseCookies = {
    cookie: {
        entr_id: string,
        entr_key: string,
        entr_hash: string
    },
    session: {
        user_id: string,
        user_name: string,
        user_hash: string,
        is_moderator: number,
        light_moderator: number,
        section: string,
        show_moderator_tools: number,
        last_error?: string   
    }
}

export const getCookies = async (): Promise<ResponseCookies> =>  {
    const json = await fetchJsonpAndGetJson(urlCookies);
    return json;
} 


// Sections
export type ResponseSection = {
    id: number,
    forum: string,
    shortn: string,
    fulln: string,
    id: number
}

export type ResponseSections = Array<ResponseSection>;

export const getSections = async (): Promise<ResponseSections> =>  {
    const json = await fetchJsonpAndGetJson(urlSections);
    return json;
} 


// New message
export type RequestNewMessage = {
    message_text: string,
    action:       "new", 
    topic_id:     string,
    rnd:          number,
    voting_select?: number,
    as_admin?: boolean    
}

export const postNewMessage = async (params: RequestNewMessage): Promise<any> =>  {

    let url = urlNewMessage.replace(':id', params.topic_id);

    await fetch(`${domain}/${url}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
    });
} 


// New topic
export type RequestNewTopic = {
    action:         "new", 
    rnd:            number,
    topic_text:     string,
    message_text:   string,
    target_section: string, 
    target_forum:   string,
    voting:         number,
    select1?:       string,    
    select2?:       string,    
    select3?:       string,    
    select4?:       string,    
    select5?:       string,    
    select6?:       string,    
    select7?:       string,    
    select8?:       string,    
    select9?:       string,    
    select10?:      string
}

export const postNewTopic = async (params: RequestNewTopic): Promise<any> =>  {

    await fetch(`${domain}/${urlNewTopic}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
        headers: {
            'Content-Type': 'application/x-www-form-urlencoded'
        }    
    });
} 


//Bookmark
export type RequestBookmark = {
    id: string
}

export const postBookmark = async (params: RequestBookmark) => {

    await fetch(`${domain}/${urlAddBookmark}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
    });
}


//Search
export type RequestSearch = {
    keywords: string
}

export type ResponseSearch = {

}

export const postSearch = async (params: RequestSearch) => {

    await fetch(`${domain}/${urlSearch}`, {
        method: 'POST',
        body: paramsToString('', params),
        mode: 'no-cors',
        credentials: 'include',
    });
}

export const paramsToString = (paramsPrefix: string, params: ?{}): string => {

    if (!params)
        return '';

    let arr = [];
    for (let key in params) {
        if (params[key] !== undefined)
            arr.push(key + '=' + params[key]);
    }

    if (arr.length > 0)
        return paramsPrefix + arr.join('&');
    else    
        return '';
}

export const fetchJsonpAndGetJson = async (url: string, params: any): Promise<any> => {

    let fullUrl = `${domain}/${url}${paramsToString('?', params)}`; 
    const response =  await fetchJsonp(fullUrl);
    let responseJson = await response.json();
    let json;
    if (typeof(responseJson) === 'string') {
        
        try {
             json = JSON.parse(responseJson)    
        } catch(e) {
            json = utils.parseJSON(responseJson);
        }
    } else {
        json = responseJson;
    }
    return json;
}
