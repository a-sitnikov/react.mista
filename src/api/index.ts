export const domain: string = localStorage.getItem('domain') || 'https://forum.mista.ru';
//export const domain = 'https://dev.mista.ru';
export const urlTopicsList: string = localStorage.getItem('urlTopicsList') || 'api/topic.php';
export const urlTopicInfo: string = localStorage.getItem('urlTopicInfo') || 'ajax_gettopic.php';
export const urlTopicMessages: string = localStorage.getItem('urlTopicMessages') || 'api/message.php';
export const urlLogin: string = localStorage.getItem('urlLogin') || 'users.php?action=do_enter';
export const urlLogout: string = localStorage.getItem('urlLogout') || 'users.php?action=exit';
export const urlCookies: string = localStorage.getItem('urlCookies') || 'ajax_cookie.php';
export const urlSections: string = localStorage.getItem('urlSections') || 'ajax_getsectionslist.php';
export const urlNewMessage: string = localStorage.getItem('urlNewMessage') || 'topic.php?id=:id&page=1';
export const urlNewTopic: string = localStorage.getItem('urlNewTopic') || 'index.php';
export const urlAddBookmark: string = localStorage.getItem('urlAddBookmark') || 'ajax_addbookmark.php';
export const urlSearch: string = localStorage.getItem('urlSearch') || 'ajax_find.php';

export * from './cookies'
export * from './login'
export * from './new_message'
export * from './newtopic'
export * from './sections'
export * from './topic_info'
export * from './topic_messages'
export * from './topics_list'
