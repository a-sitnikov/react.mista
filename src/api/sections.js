import { fetchJsonpAndGetJson, urlSections } from '.'

// Sections
export type ResponseSection = {
  id: number,
  forum: string,
  shortn: string,
  fulln: string,
  id: number
}

export type ResponseSections = Array<ResponseSection>;

export const fetchSections = async (): Promise<ResponseSections> => {
  const json = await fetchJsonpAndGetJson(urlSections);
  return json;
}
