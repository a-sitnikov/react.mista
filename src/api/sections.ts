import { fetchJsonpAndGetJson, urlSections } from '.'
import type { ISectionItem } from 'src/store'

interface IAPIResponse {
  id: number,
  forum: string,
  shortn: string,
  fulln: string
}

function convertResponse(response: IAPIResponse): ISectionItem {

  return ({
    id: response.id,
    forum: response.forum,
    code: response.shortn,
    name: response.fulln,
  })

}

export const fetchSections = async (): Promise<ISectionItem[]> => {
  const list = await fetchJsonpAndGetJson(urlSections);
  return list.map(convertResponse);
}
