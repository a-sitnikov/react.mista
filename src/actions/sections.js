//@flow
import * as API from 'src/api'
import type { ResponseSections } from 'src/api'
import type { State } from 'src/reducers'

export type REQUEST_SECTIONS = {
    type: 'REQUEST_SECTIONS',
}

export type RECEIVE_SECTIONS = {
    type: 'RECEIVE_SECTIONS',
    items: ResponseSections,
    receivedAt: Date
}

export type SectionsAction = REQUEST_SECTIONS | RECEIVE_SECTIONS;

export const requestSections = (): REQUEST_SECTIONS => ({
    type: 'REQUEST_SECTIONS'
})

export const receiveSections = (json: ResponseSections): RECEIVE_SECTIONS => ({
    type: 'RECEIVE_SECTIONS',
    items: json,
    receivedAt: new Date()
})

export const fetchSections = (): any => async (dispatch: any) => {

    let action1 = requestSections();
    dispatch(action1);

    const json: ResponseSections = await API.getSections();
    
    let action2 = receiveSections(json);
    dispatch(action2);

}

const shouldfetchSections = (state: State): boolean => {
    
    const sections = state.sections;
    
    if (!sections) 
        return true
    
    if (sections.isFetching) 
        return false
    
    if (sections.items.length > 0)
        return false;

    return true
}

export const fetchSectionsIfNeeded = (): any => (dispatch: any, getState: any) => {
    if (shouldfetchSections(getState())) {
      return dispatch(fetchSections());
    }
  }