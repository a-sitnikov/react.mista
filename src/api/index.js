//@flow
import fetchJsonp from 'fetch-jsonp'
import * as utils from 'src/utils'

export const domain: string = localStorage.getItem('domain') || 'https://forum.mista.ru';
//export const domain = 'https://dev.mista.ru';
export const urlTopicsList: string = localStorage.getItem('urlTopicsList') || 'api/topic.php';
export const urlTopicInfo: string = localStorage.getItem('urlTopicInfo') || 'ajax_gettopic.php';
export const urlTopicMessages: string = localStorage.getItem('urlTopicMessages') || 'api/message.php';
export const urlLogin: string = localStorage.getItem('urlLogin') || 'users.php?action=do_enter';
export const urlLogout: string = localStorage.getItem('urlLogout') || 'users.php?action=exit';
export const urlCookies: string = localStorage.getItem('urlCookies') || 'ajax_cookie.php';
export const urlSections: string = localStorage.getItem('urlSections') || 'ajax_getsectionslist.php';
export const urlNewMessage: string = localStorage.getItem('urlNewMessage') || 'topic.php?id=:id&page=1';
export const urlNewTopic: string = localStorage.getItem('urlNewTopic') || 'index.php';
export const urlAddBookmark: string = localStorage.getItem('urlAddBookmark') || 'ajax_addbookmark.php';
export const urlSearch: string = localStorage.getItem('urlSearch') || 'ajax_find.php';

// New message
export type RequestNewMessage = {
  message_text: string,
  action: "new",
  topic_id: string,
  rnd: number,
  voting_select?: number,
  as_admin?: boolean
}

export const postNewMessage = async (params: RequestNewMessage): Promise<any> => {

  let url = urlNewMessage.replace(':id', params.topic_id);

  let result = await fetch(`${domain}/${url}`, {
    method: 'POST',
    body: paramsToString('', params),
    mode: 'no-cors',
    credentials: 'include',
    headers: {
      'Accept': 'text/html,application/xhtml+xml,application/xml',
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    redirect: 'follow'
  });

  return result;
}


// New topic
export type RequestNewTopic = {
  action: "new",
  rnd: number,
  topic_text: string,
  message_text: string,
  target_section: string,
  target_forum: string,
  voting: number,
  select1?: string,
  select2?: string,
  select3?: string,
  select4?: string,
  select5?: string,
  select6?: string,
  select7?: string,
  select8?: string,
  select9?: string,
  select10?: string
}

export const postNewTopic = async (params: RequestNewTopic): Promise<any> => {

  await fetch(`${domain}/${urlNewTopic}`, {
    method: 'POST',
    body: paramsToString('', params),
    mode: 'no-cors',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded'
    }
  });
}


//Bookmark
export type RequestBookmark = {
  id: string
}

export const postBookmark = async (params: RequestBookmark) => {

  await fetch(`${domain}/${urlAddBookmark}`, {
    method: 'POST',
    body: paramsToString('', params),
    mode: 'no-cors',
    credentials: 'include',
  });
}

export const postSearch = async (params: RequestSearch) => {

  await fetch(`${domain}/${urlSearch}`, {
    method: 'POST',
    body: paramsToString('', params),
    mode: 'no-cors',
    credentials: 'include',
  });
}

export const paramsToString = (paramsPrefix: string, params: ?{}): string => {

  if (!params)
    return '';

  let arr = [];
  for (let key in params) {
    if (params[key] !== undefined)
      arr.push(key + '=' + params[key]);
  }

  if (arr.length > 0)
    return paramsPrefix + arr.join('&');
  else
    return '';
}

export const fetchJsonpAndGetJson = async (url: string, params: any): Promise<any> => {

  let fullUrl = `${domain}/${url}${paramsToString('?', params)}`;
  const response = await fetchJsonp(fullUrl);
  let responseJson = await response.json();
  let json;
  if (typeof (responseJson) === 'string') {

    try {
      json = JSON.parse(responseJson)
    } catch (e) {
      json = utils.parseJSON(responseJson);
    }
  } else {
    json = responseJson;
  }
  return json;
}

export const fetchAndGetJson = async (url: string, params: any, options: any): Promise<any> => {

  let fullUrl = `${domain}/${url}${paramsToString('?', params)}`;
  const response = await fetch(fullUrl, options);
  let responseJson = await response.json();
  let json;
  if (typeof (responseJson) === 'string') {

    try {
      json = JSON.parse(responseJson)
    } catch (e) {
      json = utils.parseJSON(responseJson);
    }
  } else {
    json = responseJson;
  }
  return json;
}