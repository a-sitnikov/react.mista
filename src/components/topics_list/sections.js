//@flow
import React, { Component, useEffect } from 'react'
import { connect } from 'react-redux'
import Form from 'react-bootstrap/Form'

import { getSectionsIfNeeded } from 'src/data/sections/actions'

import type { ResponseSection, ResponseSections } from 'src/api'

import type { DefaultProps } from 'src/components'
import type { State } from 'src/reducers'
import { useAppDispatch } from 'src/data/store'

type SectionSelectProps = {
  defaultValue: string,
  selected: string,
  className: string,
  id: string,
  style?: {},
  size: ?string,
  onChange: (e: any, section: ResponseSection | null) => void
}

type StateProps = {
  items: ResponseSections,
  tree: {}
}

type Props = SectionSelectProps & StateProps & DefaultProps;

const Sections = (props) => {

  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(getSectionsIfNeeded());
  }, [dispatch]);

  const onSelect: (e: SyntheticEvent<HTMLSelectElement>) => void = (e: SyntheticEvent<HTMLSelectElement>) => {

    const { items, onChange } = props;

    if (onChange) {
      const code = e.currentTarget.value;
      const arr = items.filter(val => val.code === code);
      if (arr.length > 0)
        onChange(e, arr[0]);
      else
        onChange(e, null);
    }
  }

  const { id, tree, defaultValue, selected, style, size } = props;

  let sectionsElem = [];
  for (let forum in tree) {

    let group =
      <optgroup key={forum} label={forum}>
        {tree[forum].map((item, i) => (
          <option key={item.id} value={item.code}>
            {item.name}
          </option>
        ))}
      </optgroup>

    sectionsElem.push(group);
  }

  return (
    <Form.Control as="select"
      onChange={onSelect}
      value={selected}
      style={style}
      className='input'
      size={size}
      id={id}
    >
      <option value="">{defaultValue}</option>
      {sectionsElem}
    </Form.Control>
  )
}

const mapStateToProps = (state: State): StateProps => {

  const { items, tree } = state.sections;

  return {
    items,
    tree
  }
}

export { Sections };
export default connect(mapStateToProps)(Sections);