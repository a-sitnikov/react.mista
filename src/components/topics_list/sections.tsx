import React, { FC, ReactElement, useEffect } from 'react'
import { connect, ConnectedProps } from 'react-redux'
import Form from 'react-bootstrap/Form'

import { getSectionsIfNeeded } from 'src/data/sections/actions'
import { RootState, useAppDispatch } from 'src/data/store'
import type { ISectionItem } from 'src/data/sections'

type IProps = {
  id: string,
  defaultValue: string,
  selected?: string,
  style?: {},
  size?: 'sm' | 'lg',
  onChange: (e: any, value: ISectionItem) => void
}

const mapState = (state: RootState) => {

  const { items, tree } = state.sections;

  return {
    items,
    tree
  }
}

const connector = connect(mapState);
const Sections: FC<ConnectedProps<typeof connector> & IProps> = (props): ReactElement => {

  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(getSectionsIfNeeded());
  }, [dispatch]);

  const onSelect: (e: React.ChangeEvent<HTMLElement>) => void = (e: any) => {

    const { items, onChange } = props;

    if (onChange) {
      const code = e.currentTarget.value;
      const arr = items.filter(item => item.code === code);
      if (arr.length > 0)
        onChange(e, arr[0]);
      else
        onChange(e, null);
    }
  }

  const { id, tree, defaultValue, selected, style, size } = props;

  let sectionsElem = [];
  for (let forum in tree) {

    let group =
      <optgroup key={forum} label={forum}>
        {tree[forum].map((item) => (
          <option key={item.id} value={item.code}>
            {item.name}
          </option>
        ))}
      </optgroup>

    sectionsElem.push(group);
  }

  return (
    <Form.Select
      onChange={onSelect}
      value={selected}
      style={style}
      className='input'
      size={size}
      id={id}
    >
      <option value="">{defaultValue}</option>
      {sectionsElem}
    </Form.Select>
  )
}

export { Sections };
export default connector(Sections);