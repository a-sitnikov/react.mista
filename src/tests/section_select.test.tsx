import { screen, waitFor } from "@testing-library/react"
import { renderWithProviders } from './test-utils'

import * as API from 'src/api/sections'
import SectionSelect from 'src/pages/topics_list/sections'
import type { IProps } from 'src/pages/topics_list/sections'
import { ISectionItem } from 'src/store';

const data: ISectionItem[] = [
    {
        "forum": "1C",
        "code": "v7",
        "name": "1С:Предприятие 7.7 и ранее",
        "id": 3
    },
    {
        "forum": "1C",
        "code": "v8",
        "name": "1С:Предприятие 8 общая",
        "id": 8
    }
];

describe('SectionSelect', ()=>{

    it('render', async () => {
        const props: IProps = {
            defaultValue: 'Секция',
            selected: '',
            id: 's1',
            size: 'sm'
        }
        
        jest.spyOn(API, "fetchSections").mockImplementation(async () => Promise.resolve(data))

        renderWithProviders(
            <SectionSelect {...props} />
        );
        await waitFor(() => expect(API.fetchSections).toHaveBeenCalledTimes(1));
        
        const options = screen.getAllByRole('option');
        expect(options.length).toBe(3);
        expect(options[0].textContent).toBe('Секция');
        expect(options[1].textContent).toBe(data[0].name);
        expect(options[2].textContent).toBe(data[1].name);

    });

})    