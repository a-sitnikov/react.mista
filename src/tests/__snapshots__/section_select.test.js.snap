// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionSelect +++ render the connected() component 1`] = `"<select id=\\"s1\\" class=\\"input form-control form-control-sm\\"><option value=\\"\\">Секция</option><optgroup label=\\"1C\\"><option value=\\"v7\\">1С:Предприятие 7.7 и ранее</option><option value=\\"v8\\">1С:Предприятие 8 общая</option></optgroup></select>"`;
